Transition to Apache License 2.0: License Update, Author Attribution, and Documentation Compliance

vschaffn opened the pull request:
# Description

This PR implements the transition of the **xDEM** project license from **MIT** to **Apache License 2.0**.
Resolves #555 

# Changes implemented
- The `LICENSE` file has been updated with the full text of the Apache v2.0 license.
- A `NOTICE` file has been added to comply with Apache licensing requirements. The file includes necessary copyright notices and details about dependencies used by xdem, including Python, NumPy, Matplotlib, SciPy, Geoutils, Rasterio, GeoPandas, Numba, pyproj, tqdm, scikit-image, scikit-gstat, pyogrio, pandas, scikit-learn, OpenCV, affine, Shapely, pytransform3d, yaml and noisyopt.
- An `AUTHORS.md` file has been created listing the main contributors of the project along with their contact information, while mentioning their affiliation with the **xDEM** developers' copyright.
- All source code files have been updated with the new Apache v2.0 license header, which is the following:

```
# Copyright (c) 2024 xDEM developers.
#
# This file is part of the xdem project:
#   https://github.com/glaciohack/xdem
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
```
The header is set up automatically via pre-commit for every source files.

- The `setup.cfg` has been updated with the new license and copyright as well as `conf.py`.

- The documentation has been updated with the license and the authors.

# Notes

- There are certainly other contributors to add to the project, so don't hesitate to let me know.
- The packages listed in `NOTICE` and in `setup.cfg` are not aligned anymore. This will be solved by #582.
- PR to be merged after #612.
- The copyrights and authors cited in the documentation must be checked for consistency with the project #431.

adebardo commented on AUTHORS.md:
```diff
@@ -0,0 +1,32 @@
+- **Ferdinand Schenck** [@fnands](https://github/fnands)
+- **Johannes Landmann** [@jlandmann](https://github/jlandmann)
+
+Update here with new contributors.
```
Could you redesign this file to make it look more like this one https://github.com/CNES/Pandora2D/blob/master/AUTHORS.md?

adebardo commented on LICENSE:
```diff
@@ -1,21 +1,176 @@
+      incurred by, or claims asserted against, such Contributor by reason
+      of your accepting any such warranty or additional liability.
+
+   END OF TERMS AND CONDITIONS
```
"Isn't the appendix missing?"

adebardo commented on NOTICE:
```diff
@@ -0,0 +1,79 @@
+scikit-gstat: A geostatistics toolbox for Python.
+Copyright (c) 2018-2023 scikit-gstat developers.
+Website: https://github.com/mmaelicke/scikit-gstat
+Licence: MIT License.
```
Modules Affine, Pandas, Pyogrio, SciKit Learn, Shapely,  OpenCV packages for Python, Packaging are missing. 
I know they are missing in the requirements file, you can do it now or wait for the developments related to the makefile.

adebardo commented on setup.cfg:
```diff
@@ -6,15 +6,15 @@ description = Analysis of digital elevation models (DEMs)
 keywords = dem, elevation, geoutils, xarray
```
In demcompare, the setup files have a header; I'm checking to see if it's necessary here or not.

duboise-cnes replied:
not needed, doesn't need to be overcopyrighted ;) setup follows usually a standard way and no invention to be copyrighted.

adebardo commented on setup.cfg:
```diff
@@ -6,15 +6,15 @@ description = Analysis of digital elevation models (DEMs)
     Intended Audience :: Science/Research
     Natural Language :: English
-    License :: OSI Approved :: MIT License
+    License :: OSI Approved :: Apache License 2.0
```
Here it is "Apache Software License"

adebardo commented on xdem/__init__.py:
```diff
@@ -1,3 +1,20 @@
+# Copyright (c) 2024 xdem developers
```
I'm checking on the dates because I don't think 2024 is the correct one.

adebardo commented:
I'm looking into setting up a changelog file.

rhugonnet commented:
Really great to have this done rigorously! :slightly_smiling_face:  
On the next step you list for adding more contributors: I would start obviously by adding yourselves @vschaffn @adebardo :wink:. Then, should we consider listing those that do not have commits but contributed to important amount of code through other packages that have been or are being adapted in xDEM? (@duboise-cnes, but also for instance @iamdonovan & potentially others; if they feel it is warranted).

Other than that, I have three remarks:
1. Should we also modify the "author" name of PyPI with the same "xdem contributors" name? https://github.com/GlacioHack/xdem/blob/3287594522d91542a6b747a77c931597f4db28ad/setup.cfg#L2
2. Should we also update the documentation copyright here: https://github.com/GlacioHack/xdem/blob/3287594522d91542a6b747a77c931597f4db28ad/doc/source/conf.py#L28 
I think I remember @sdinot mentioning the copyright for the documentation being separate? What would be the best practice here? (it seems easier to have a single one)
3. Should we also update the Zenodo repository of the package accordingly? https://zenodo.org/records/4809698.

Final note: if we address these points consistently, this PR would also resolve #431!

rhugonnet commented on AUTHORS.md:
```diff
@@ -0,0 +1,45 @@
+
+---
+
+- **Romain Hugonnet** [@rhugonnet](https://github.com/rhugonnet) <hugonnet@uw.edu>
```
Is it common to have the emails of the main developers in this file? Just out of curiosity. If yes, OK for me!
We can replace by my public GitHub email instead: romain.hugonnet@gmail.com (as our institutional emails usually become deprecated after a couple years in early research years, moving all the time :sweat_smile: !)

rhugonnet replied:
Amaury's email should remain the same for a while though!

duboise-cnes replied:
I don't know if it is common, but we have done that to be able to find all contributors if one day there is a need to contact all contributors on a global projet matter (communication, licence, global info...). So I think it is a good habit to have.

adehecq replied:
Just looking into this a posteriori. I would prefer if the email addresses are not used because 1) they may evolve in the future and I'm sure they will not be rigorously updated 2) it's almost an incentive for spamming us :wink:.
I would suggest pointing to the GitHub pages instead. Or I think GitHub can create aliases for emails no?
What do you think @rhugonnet @duboise-cnes @adebardo ?

rhugonnet commented on AUTHORS.md:
```diff
@@ -0,0 +1,45 @@
+# Credits
+
+---
+© 2024 **xdem developers**.
```
If this applies here: For use in plain text purposes, we had previously converged on the writing "xDEM" (currently consistent nearly everywhere in documentation, README, etc).

Happy to re-discuss that syntax if you have an opinion! @duboise-cnes @adebardo @vschaffn
Also tagging @adehecq @erikmannerfelt who were part of the initial discussion.
In short our logic was:
1. We want to have upper-case where relevant in the name, same as "NumPy", "SciPy", etc (not the case for all package, like for "pandas", but usually is),
2. We chose to keep the DEM upper-case to clarify its acronym origin.

The advantage of that writing is that makes it obvious that it's about DEMs (which was our priority), while some disadvantages are that it starts with a lower-case (slightly atypical) and is not consistent with the writing of "Xarray" (not a big worry).

Not a super important topic, it's more a question of style, but we might as well get it right directly in the new License with everyone onboard! :slightly_smiling_face:

duboise-cnes replied:
as you see fit. xDEM everywhere perfect for me.

sdinot commented:
> 2\. I think I remember @sdinot mentioning the copyright for the documentation being separate? What would be the best practice here? (it seems easier to have a single one)

To be precise, I wasn't talking about the copyright notice, but about the license, which must be adapted to the nature of the protected work. The Apache v2.0 license is a permissive open source license adapted to software. Its vocabulary and clauses make it less suitable for documentation. For the latter, a permissive open source license dedicated to literary and artistic works, such as the [Creative Commons Attribution 4.0 International](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0) license, would be more appropriate.

Those of you who read French (or have a good automatic translator) can read the blog post I wrote on this subject:
[Projet libre : à chaque ressource sa licence](https://www.palabritudes.net/2021/05/05/projet-libre-a-chaque-ressource-sa-licence.html).

rhugonnet commented:
Thanks @sdinot, your blog post is a great resource! (I'll also peak at the /e/OS/ post for my Fairphone later :smile:)
Adding a CC BY 4.0 for the documentation sounds perfect, and potentially a 0BSD for code snippets/examples.

You give an example of what the package's README would look like. Would it also be good to add a short version of this notice at the bottom of the documentation pages? If yes, what would this look like (something like: "Copyright xDEM developers: text and figures CC BY 4.0, code snippets 0BSD")?

duboise-cnes commented:
> Really great to have this done rigorously! 🙂 On the next step you list for adding more contributors: I would start obviously by adding yourselves @vschaffn @adebardo 😉. Then, should we consider listing those that do not have commits but contributed to important amount of code through other packages that have been or are being adapted in xDEM? (@duboise-cnes, but also for instance @iamdonovan & potentially others; if they feel it is warranted).
> 
> Other than that, I have three remarks:
> 
>     1. Should we also modify the "author" name of PyPI with the same "xdem contributors" name? https://github.com/GlacioHack/xdem/blob/3287594522d91542a6b747a77c931597f4db28ad/setup.cfg#L2
> 
>     2. Should we also update the documentation copyright here: https://github.com/GlacioHack/xdem/blob/3287594522d91542a6b747a77c931597f4db28ad/doc/source/conf.py#L28
>        
>        I think I remember @sdinot mentioning the copyright for the documentation being separate? What would be the best practice here? (it seems easier to have a single one)
> 
>     3. Should we also update the Zenodo repository of the package accordingly? https://zenodo.org/records/4809698.
> 
> 
> Final note: if we address these points consistently, this PR would also resolve #431!

I am ok to be added in contributors list (with emmanuel.dubois@cnes.fr) even no commits yet (I will try at some point when less crowded in CNES) : that allows to identify CNES in contributions simply. 
Don't forget to add CNES copyright on "real" further contributions on future code from demcompare (CLI, ....). Not needed on this PR, which is only a copyright clean, an important but not complex contribution. So ok  for this one, but to be careful on following ones to choose wisely where to go on copyright. If it is not clear, ask me PR after PR on CNES funded ones.

Pypi author and documentation should be also changed, i agree.
For the documentation, see sebastien answer but not needed on each file but globally with any adapted documentation licence. Great if it is clean but doesn't have the best choice, simply an understable and classical ones for readers and doc contributors. 

Hope it helps to finish this PR

vschaffn commented:
Many thanks for your feedback !
@adebardo Every packages used in source are now listed in `NOTICE`. Then I will modify setup.cfg but this would be for #582.
@rhugonnet I have changed `xdem` to `xDEM` wherever applicable, and added the contributors you mentionned.
@duboise-cnes I have noted your comment about the CNES copyright, and I will keep that in mind for future contributions.

rhugonnet commented on AUTHORS.md:
```diff
@@ -0,0 +1,49 @@
+
+- **Romain Hugonnet** [@rhugonnet](https://github.com/rhugonnet) <romain.hugonnet@gmail.com>
+- **Amaury Dehecq** [@adehecq](https://github/adehecq) <amaury.dehecq@univ-grenoble-alpes.fr>
+- **Erik Schytt Mannerfelt** [@erikmannerfelt](https://github/erikmannerfelt)
```
@erikmannerfelt Do you want an email to be added for you here?

rhugonnet replied:
@erikmannerfelt Reminder on this

rhugonnet commented on AUTHORS.md:
```diff
@@ -0,0 +1,49 @@
+
+---
+
+- **Romain Hugonnet** [@rhugonnet](https://github.com/rhugonnet) <hugonnet@uw.edu>
```
Change email here too!

rhugonnet commented on AUTHORS.md:
```diff
@@ -0,0 +1,49 @@
+- **Alessandro Gentilini** [@alessandro-gentilini](https://github/alessandro-gentilini)
+- **Ferdinand Schenck** [@fnands](https://github/fnands)
+- **Johannes Landmann** [@jlandmann](https://github/jlandmann)
+- **Valentin Schaffner** [@vschaffn](https://github/vschaffn) <valentin.schaffner@cs-soprasteria.com>
```
Need to add @MatteaE as a contributor here as well (from the list of committers).
And also @ameliefroessl for all her work on Dask in https://github.com/GlacioHack/xdem/discussions/501 and #525, even if this has been on pause since the summer (sorry, too much to do this fall!) :wink:.

Other actors I'm tagging that can tell us if they want to be added:
@fmaussion: You helped a lot starting up the package during the first hackweek, advising on good practices for CI/doc/tests and sharing code for that (in addition to discussing about the landscape for Xarray geospatial). You have commits on GeoUtils but not here, I still feel like you should be added as contributor. What do you think?
@sgascoin: For providing a base algorithm + tips on the optimizer for a new coregistration method, if you feel it is warranted.

sgascoin replied:
Sure, I'm happy to be in the friends list!

rhugonnet replied:
@fmaussion Reminder on this

rhugonnet commented:
All good for me, just the couple comments above! 
We can add the contributors I asked later (in a separate PR), so no problem to merge this directly next week :slightly_smiling_face:.

rhugonnet commented on AUTHORS.md:
```diff
@@ -0,0 +1,49 @@
+contributions are subject to the project's copyright under the terms of the
+[Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0).
+
+See CONTRIBUTING.md for more details on Contributor License Agreement.
```
Given this statement, do we need to have a "Contributor License Agreement" section in `CONTRIBUTING.md`, or is this just a generic terminology?
(We can also add "the" before "Contributor License Agreement"!)

duboise-cnes replied:
Good review @rhugonnet : We can indeed remove this statement. We have chosen not to go to a contributor license agreement model. If needed at some point, this will be further evolution.

adebardo replied:
@vschaffn we can remove this statement

sdinot commented:
> You give an example of what the package's README would look like. Would it also be good to add a short version of this notice at the bottom of the documentation pages? If yes, what would this look like (something like: "Copyright xDEM developers: text and figures CC BY 4.0, code snippets 0BSD")?

We could indeed consider placing a copyright notice on all pages of the documentation, as we do in the source code. But the practice is different for documentation. In general, authors simply indicate the license of the code, documentation and other project resources on a dedicated page of the documentation.

This difference in practice between source code and documentation is probably due to the fact that a source code file can more easily be extracted from its context and reused in another than a page of documentation. We therefore ensure that each source code file is legally “self-supporting”. It contains not only the source code, but also the name of the copyright holder, the license and the origin of the file.

duboise-cnes commented on AUTHORS.md:
```diff
@@ -0,0 +1,52 @@
+- **Johannes Landmann** [@jlandmann](https://github/jlandmann)
+- **Valentin Schaffner** [@vschaffn](https://github/vschaffn) <valentin.schaffner@cs-soprasteria.com>
+- **Alice De Bardonnèche-Richard** [@adebardo](https://github/adebardo) <alice.de-bardonneche-richard@cs-soprasteria.com>
+- **Emmanuel Dubois** [@dubois-cnes](https://github/dubois-cnes) <emmanuel.dubois@cnes.fr>
```
@duboise-cnes (lacks an "e"

rhugonnet commented:
@vschaffn I re-tagged @erikmannerfelt and @fmaussion regarding the author listing + email.
If they don't answer in the next days, feel free to merge the PR anyways, we can adjust those things later on :wink:.

adebardo commented on xdem/coreg/base.py:
```diff
@@ -1,3 +1,20 @@
+# Copyright (c) 2024 xDEM developers
+#
+# This file is part of xdem project:
```
There are still "xdem" placeholders in the headers.

adebardo commented on AUTHORS.md:
```diff
@@ -0,0 +1,52 @@
+
+- **Romain Hugonnet** [@rhugonnet](https://github.com/rhugonnet) <romain.hugonnet@gmail.com>
+- **Amaury Dehecq** [@adehecq](https://github/adehecq) <amaury.dehecq@univ-grenoble-alpes.fr>
+- **Erik Schytt Mannerfelt** [@erikmannerfelt](https://github/erikmannerfelt)
```
@erikmannerfelt Hello, I’m tagging you because we are about to merge this PR that will change the license of xDEM. Would you like us to include your email?

adebardo reviewed (commented):
@duboise-cnes Could you please indicate how we can change the information on PyPI? 

@vschaffn Could you add a file of the license and authors in the documentation ? as an example : https://www.orekit.org/site-orekit-development/licenses.html

@sdinot said "In general, authors simply indicate the license of the code, documentation and other project resources on a dedicated page of the documentation"

vschaffn commented:
@adebardo @rhugonnet I updated the PR following your feadbacks, the license and the authors are now in the documentation.
Moreover, I have set up header automation for the source code, and fixed a minor issue with make clean.

rhugonnet commented on doc/source/index.md:
```diff
@@ -134,6 +134,8 @@ advanced_examples/index.rst
 api
+authors
+license
 config
```
My only comment would be to move the Authors + License pages at the very end of the list here. (to have Configuration/Release notes pages showing before, as it seems more important for users)

We could also think about merging them together and potentially into sub-sections of the "Background" page instead of each being their own page, or into a new section "Project information" that includes "Publications", "License", "Authors", etc... (to avoid mixing these with the API), but that can be discussed/done later! :slightly_smiling_face:

vschaffn replied:
@rhugonnet  for now I added a new section "Project information" at the end of the documentation, that includes "Publications", "Background", "Authors" and "License".

rhugonnet replied:
Perfect!

rhugonnet commented:
Just a small comment above, then yes all good! :smile:

adehecq commented on NOTICE:
```diff
@@ -0,0 +1,127 @@
+LICENSE file or http://www.apache.org/licenses/LICENSE-2.0 for details.
+
+Python: programming language that lets you work quickly and integrate systems more effectively.
+Copyright (c) 2001-2023 Python Software Foundation. All Rights reserved.
```
Just out of curiosity, will we have to update the dates annually in this file? @adebardo

adehecq commented:
Thanks for the meticulous work @vschaffn @adebardo !

Should we apply the same changes to GeoUtils at some point? @rhugonnet 
I guess it is not urgent as the current BSD license is permissive, but would be good to be consistent...
